code/modules/power/chem_combustion.dm

```diff
@@ -0,0 +1,653 @@
+			return
+
+		else
+			src.post_status(target_id, "command", "status", "data", "anchored=[src.anchored]&fuel_rating=[src.last_fuel]&oxygen=[src.last_oxygen * 100]%&power=[src.last_output] W&inlet=[(src.last_inlet / 2) * 100]%&mix=[src.last_mix]", "vars", "active=1&fuel_inlet=[fuel_inlet * 100]%&air_inlet=[atmos_inlet * 100]%", "device", src.device, multicast = 1)
+			return
+
+	proc/post_status(var/target_id, var/key, var/value, var/key2, var/value2, var/key3, var/value3, var/key4, var/value4, var/multicast)
+		if ((!target_id && !multicast) || !src.packet_control)
+			return
+
+		var/datum/signal/signal = get_free_signal()
+		signal.source = src
+		signal.data["sender"] = src.net_id
+
+		signal.data[key] = value
+		if (target_id)
+			signal.data["address_1"] = target_id
+		if (multicast)
+			signal.data["address_tag"] = ADDRESS_TAG_POWER
+		if (key2)
+			signal.data[key2] = value2
+		if (key3)
+			signal.data[key3] = value3
+		if (key4)
+			signal.data[key4] = value4
+
+		SEND_SIGNAL(src, COMSIG_MOVABLE_POST_RADIO_PACKET, signal)
+
+	proc/get_fuel_power(datum/reagents/R)
+		if (!R || !R.total_volume)
+			return
+
+		var/average = 0
+		for (var/reagent_id in R.reagent_list)
+			var/datum/reagent/current_reagent = R.reagent_list[reagent_id]
+
+			if (reagent_id in src.valid_fuels)
+				average += src.valid_fuels[reagent_id] * current_reagent.volume
+
+		if (!average)
+			return FALSE
+
+		return average / R.total_volume
+
+	proc/start_engine(var/status)
+		if (!src.active)
+			if (!src.ready_to_start())
+				return TRUE
+
+			src.active = TRUE
+			src.UpdateIcon()
+			src.updateIntDialog()
+			if (status)
+				src.send_status()
+
+			if (!ON_COOLDOWN(src, "tractor", 2 SECOND))
+				playsound(src.loc, 'sound/machines/tractorrev.ogg', 40, pitch=2)
+
+			return FALSE
+
+	proc/stop_engine(var/status = TRUE)
+		if (src.active)
+			src.active = FALSE
+			src.UpdateIcon()
+			src.updateIntDialog()
+			if (status)
+				src.send_status()
+
+			if (!ON_COOLDOWN(src, "tractor", 2 SECOND))
+				playsound(src.loc, 'sound/machines/tractorrev.ogg', 40, pitch=2)
+
+			return FALSE
+
+	proc/ready_to_start()
+		if (!anchored || !src.fuel_tank)
+			return FALSE
+
+		if (!src.get_fuel_power(src.fuel_tank.reagents) || !src.check_available_oxygen())
+			return FALSE
+
+		return TRUE
+
+	// Returns the concentration of oxygen in the available gas_mixture
+	proc/check_available_oxygen()
+		if (src.inlet_tank)
+			return src.check_tank_oxygen(src.inlet_tank)
+
+		var/turf/simulated/T = get_turf(src)
+		if (!istype(T))
+			if (!istype(T, /turf/))
+				return
+
+			var/datum/gas_mixture/G = T.return_air()
+			if (G.oxygen)
+				return T.oxygen / TOTAL_MOLES(G)
+
+			else
+				return FALSE
+
+		if (!T.air || T.air.oxygen <= 0)
+			return FALSE
+
+		return T.air.oxygen / TOTAL_MOLES(T.air)
```

Comment:
This is still incorrect if `parent.group_processing == TRUE`

Should be using `return_air()`